### DISCUSSION TITLE
How do I add more information while parsing?

### DISCUSSION THREAD

**Discussion by awson** [foonathan/lexy#104] [Q&A] [answered]
Let's we want to parse numbers with lexy, but not convert them to numbers proper, only parse as a string, like, e.g., in

```cpp
struct integer : lexy::token_production
{
	static constexpr auto rule = [] {
		auto hex_integer = (dsl::lit<"0x"> | dsl::lit<"0X">) >> dsl::capture(dsl::token(dsl::while_(dsl::digit<> / dsl::ascii::one_of<"abcdefABCDEF">)));
		auto regular_integer = dsl::peek(dsl::ascii::one_of<"-+"> / dsl::digit<>) >> dsl::capture(dsl::token(dsl::while_(dsl::digit<>)));
		return hex_integer | regular_integer;
	}();
	static constexpr auto value = lexy::as_string<std::u8string_view>;
};
```

Thus we get a string view into the parsed number, but we lose the information about the radix of the number parsed.

I'd want to parse the number into something like

```cpp
struct number {
  std::u8string_view num_lit;
  int num_radix;
};
```

But, since we capture in both cases, it seems it's impossible to to use a custom callback, since it can't discriminate between the cases.

Thus, the question is what is the best way to accomplish the task with idiomatic lexy?

**Comment by foonathan** [comment, marked as answer]
First, `dsl::digit<> / dsl::ascii::one_of<"abcdefABCDEF">` is `dsl::digit<dsl::hex>`, and `dsl::while_one(dsl::digit<T>)` is `dsl::digits<T>`: https://lexy.foonathan.net/reference/dsl/digit/ Second, if you have `condition >> a | dsl::peek(condition) >> b`, you can also use `condition >> a | dsl::else_ >> b`.

Then to answer your main question: to produce a value you need to parse a rule that produces a value. The recommended way to achieve it is the use of a production. Overall, I'd write it like so:

```cpp
struct integer : lexy::token_production
{
    struct hex_prefix
    {
        static constexpr auto rule = dsl::lit<"0x"> | dsl::lit<"0X">;
        static constexpr auto value = lexy::constant(16);
    };
    
    struct decimal_prefix
    {
        static constexpr auto rule = dsl::lit(""); // parse nothing
        static constexpr auto value = lexy::constant(10);
    };

	static constexpr auto rule = [] {
		auto hex_integer = dsl::p<hex_prefix> >> dsl::capture(dsl::digits<dsl::hex>);
		auto regular_integer = dsl::p<decimal_prefix> + dsl::capture(dsl::digits<>);
		return hex_integer | dsl::else_ >> regular_integer;
	}();
	static constexpr auto value = lexy::callback([](int base, auto digits) {
	    …
	});
};
```

**Reply by awson** [reply]
Ah, thanks for the very useful reply.

I'll try to recreate the question, but not right now.

I've deleted the question since, right after publishing it, I immediately recognized that I can create two completely separate productions for `hex_integer` and `regular_integer` which both are valued as `struct number` (i.e. string_view literal value + int radix value).

**Reply by foonathan** [reply]
> I'll try to recreate the question, but not right now.

I've realized that I am also able to edit your post, so I've restored it.

**Reply by awson** [reply]
Thanks!

**Reply by awson** [reply]
I've tried your suggestion (with prefixes) and got:

1. Zero-size array warning.
2. A **completely different rule starts failing** with `static_assert failed: 'choice requires a branch condition'`. 

My code is derived from your `calculator` example, and the failing rule is `var_or_call` (BUT generalized to arbitrary number of arguments).

If I use 2 completely separate productions , things go perfectly:
```
struct hex_integer : lexy::token_production {
	rule_ (dsl::lit<"0x"> | dsl::lit<"0X">) >> dsl::capture(dsl::digits<dsl::hex>);
	value_ lexy::as_string<string_view> | lexy::callback<ast::num_lit>([](string_view n) {return ast::num_lit(n, 16);});
};

struct reg_integer : lexy::token_production {
	rule_ dsl::capture(dsl::digits<>);
	value_ lexy::as_string<string_view> | lexy::callback<ast::num_lit>([](string_view n) {return ast::num_lit(n, 10);});
};

...

auto num_literal = dsl::p<hex_integer> | dsl::p<reg_integer>;
```
(`string_view` is typdefed as `std::u8string_view`, `ast::num_lit` is the `struct number` from the question, `rule_`  and `value_` are abbreviation macros for `static constexpr auto xxxx =`).

**Reply by foonathan** [reply]
Ah, yes. When you use `dsl::else_`, `integer` itself is no longer a branch rule (since the choice is always unconditionally taken due to the else). So it does make sense to use peek here (or just `digits` directly as that's also a branch rule).
